reactor-core/src/main/java21/reactor/core/scheduler/LoomBoundedElasticScheduler.java

```diff
@@ -0,0 +1,1352 @@
+			Objects.requireNonNull(unit, "TimeUnit should be non-null");
+
+			if (isShutdown(this.state)) {
+				throw Exceptions.failWithRejected();
+			}
+
+			ensureQueueCapacityAndAddTasks(1);
+
+			Runnable decoratedTask = onSchedule(task);
+			boolean isDirect = disposables == null;
+			SchedulerTask disposable =
+					new SchedulerTask(this, decoratedTask, -1, unit, disposables);
+
+			if (!isDirect) {
+				if (!disposables.add(disposable)) {
+					throw Exceptions.failWithRejected();
+				}
+			}
+
+			if (delay <= 0) {
+				this.tasksQueue.offer(disposable);
+
+				trySchedule();
+			}
+			else {
+				disposable.schedule(delay, unit);
+			}
+
+			return disposable;
+		}
+
+		Disposable schedulePeriodically(Runnable task,
+				long initialDelay,
+				long period,
+				TimeUnit unit,
+				@Nullable Composite disposables) {
+			Objects.requireNonNull(unit, "TimeUnit should be non-null");
+
+			if (isShutdown(this.state)) {
+				throw Exceptions.failWithRejected();
+			}
+
+			ensureQueueCapacityAndAddTasks(1);
+
+			Runnable decoratedTask = onSchedule(task);
+			boolean isDirect = disposables == null;
+
+			SchedulerTask disposable = new SchedulerTask(this,
+					decoratedTask,
+					period < 0 ? 0 : period,
+					unit,
+					disposables);
+
+			if (!isDirect) {
+				if (!disposables.add(disposable)) {
+					throw Exceptions.failWithRejected();
+				}
+			}
+
+			if (initialDelay <= 0) {
+				this.tasksQueue.offer(disposable);
+
+				trySchedule();
+			}
+			else {
+				disposable.schedule(initialDelay, unit);
+			}
+
+			return disposable;
+		}
+
+		void trySchedule() {
+			int previousState = addWork(this);
+			if (previousState != 0) {
+				return;
+			}
+
+			drain();
+		}
+
+		void tryForceSchedule() {
+			int previousState = forceAddWork(this);
+			if (previousState != 0) {
+				return;
+			}
+
+			drain();
+		}
+
+		void drain() {
+			final Queue<SchedulerTask> q = this.tasksQueue;
+
+			int m = this.state;
+			for (;;) {
+				for (;;) {
+					if (isDisposedNow(this.markCount)) {
+						clearAllTask();
+						countDown();
+						return;
+					}
+
+					final SchedulerTask task = q.poll();
+
+					if (task == null) {
+						break;
+					}
+
+					removeTask();
+
+					this.activeTask = task;
+					if (task.start()) {
+						return;
+					}
+				}
+
+				m = markWorkDone(this, m);
+
+				if (isShutdown(m)) {
+					countDown();
+					return;
+				}
+
+				if (!hasWork(m)) {
+					return;
+				}
+			}
+		}
+
+		void clearAllTask() {
+			SchedulerTask activeTask = this.activeTask;
+			if (activeTask != null) {
+				activeTask.dispose();
+			}
+
+			Queue<SchedulerTask> q = this.tasksQueue;
+
+			SchedulerTask d;
+			while ((d = q.poll()) != null) {
+				d.dispose();
+			}
+		}
+
+		/**
+		 * Is this {@link BoundedState} still in use by workers.
+		 *
+		 * @return true if in use, false if it has been disposed enough times
+		 */
+		@Override
+		public boolean isDisposed() {
+			return isDisposed(this.markCount);
```

Comment:
the argument should be the state, not markCount, or I am missing something?

Reply:
let me revisit that